<a href="https://colab.research.google.com/github/data602sps/datasetspractice/blob/main/07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section. 

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html 

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

I'm going to choose this dataset on NYC OpenData called (New York City Leading Causes of Death)[https://data.cityofnewyork.us/Health/New-York-City-Leading-Causes-of-Death/jb7j-dtam/about_data] the reason is because it was released from the Department of Health and Mental Hygiene and I'm curious how they even report touchy subjects like this. It's also not a massive dataset, which I've been using a lot and it's been annoying to try and download/run locally.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include: 

* Summary statistics means, medians, quartiles, 
* Missing value information
* Any other relevant information about the dataset.  



In [ ]:
import pandas as pd

url = "https://data.cityofnewyork.us/resource/jb7j-dtam.csv"

death_df = pd.read_csv(url)

death_df.info()
death_df.head()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     1000 non-null   int64  
 1   leading_cause            1000 non-null   str    
 2   sex                      1000 non-null   str    
 3   race_ethnicity           1000 non-null   str    
 4   deaths                   1000 non-null   int64  
 5   death_rate               664 non-null    float64
 6   age_adjusted_death_rate  664 non-null    float64
dtypes: float64(2), int64(2), str(3)
memory usage: 121.3 KB


,year,leading_cause,sex,race_ethnicity,deaths,death_rate,age_adjusted_death_rate
0,2021,All Other Causes,Male,Hispanic,1695,141.8,162.7
1,2021,Malignant Neoplasms (Cancer: C00-C97),Female,Non-Hispanic White,2427,177.6,112.8
2,2021,Influenza (Flu) and Pneumonia (J09-J18),Male,Non-Hispanic Black,196,23.2,22.5
3,2021,Covid-19,Female,Hispanic,966,76.1,66.8
4,2021,"Accidents Except Drug Poisoning (V01-X39, X43,...",Male,Non-Hispanic White,266,20.1,16.6


Just viewing the dataset and the columns with a couple of examples via head(). 

In [19]:
cat_cols = ["year", "leading_cause", "sex", "race_ethnicity"]
num_cols = ["deaths", "death_rate", "age_adjusted_death_rate"]

# Summary Statistics on num cols
death_df[num_cols].describe()



,deaths,death_rate,age_adjusted_death_rate
count,1000.000000,664.000000,664.000000
mean,417.928000,59.341386,53.832673
std,776.919628,74.566312,67.116102
min,1.000000,3.557579,3.369561
25%,16.000000,13.900000,12.619500
50%,127.500000,21.615178,19.518790
75%,341.250000,90.305183,86.281513
max,4518.000000,378.200000,414.594473


In [20]:

death_df[num_cols].isna().sum()

deaths                       0
death_rate                 336
age_adjusted_death_rate    336
dtype: int64

Looks like the *death_rate* and *age_adjusted_death_rate* both has 336 NAs.

In [25]:
cat_cols = ["year", "leading_cause", "sex", "race_ethnicity"]

unique_dfs = []

for col in cat_cols:
    unique_df = (
        death_df[col]
        .drop_duplicates()
        .sort_values()
        .reset_index(drop=True)
        .to_frame(name=col)
    )
    
    unique_dfs.append(unique_df)

for i in range(len(unique_dfs)):
    print(unique_dfs[i])


   year
0  2015
1  2016
2  2017
3  2018
4  2019
5  2020
6  2021
                                        leading_cause
0   Accidents Except Drug Poisoning (V01-X39, X43,...
1                                    All Other Causes
2                           Alzheimer's Disease (G30)
3                                   Anemias (D50-D64)
4                Aortic Aneurysm and Dissection (I71)
5         Assault (Homicide: U01-U02, Y87.1, X85-Y09)
6                  Assault (Homicide: Y87.1, X85-Y09)
7                               Atherosclerosis (I70)
8           Cerebrovascular Disease (Stroke: I60-I69)
9   Certain Conditions Originating in the Perinata...
10  Certain Conditions originating in the Perinata...
11  Cholelithiasis and Disorders of Gallbladder (K...
12     Chronic Liver Disease and Cirrhosis (K70, K73)
13  Chronic Liver Disease and Cirrhosis (K70, K73-...
14               Chronic Liver Diseases and Cirrhosis
15       Chronic Lower Respiratory Diseases (J40-J47)
16  Complications 

In [26]:
death_df[cat_cols].isna().sum()

year              0
leading_cause     0
sex               0
race_ethnicity    0
dtype: int64

Looks like there's a data mix with *sex* column has inconsistent values for both Male and Female (need to fix). However, there doesn't seem to be any NAs. 

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables. 

8. Filter your data based on some condition. 

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group. 
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups. 

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

## Data Wrangling

cat_cols = ["year", "leading_cause", "sex", "race_ethnicity"]
num_cols = ["deaths", "death_rate", "age_adjusted_death_rate"]

**Modify multiple column names**

I think the column names are fine, however we could alter *leading_cause* -> *cause* and *race_ethnicity* to *race*, and *age_adjusted_death_rate* to *age_adj_death_rate*

**Fix missing and invalid values in data**

We can fix the *sex* column to make `M -> Male` and `F -> Female`, that would fix the column's issue of multiple codes for a specific sex. 

**Create new columns**

We can create a new column called *adj_rate* which will show how much age adjusts the death_rate for *age_adj_death_rate*

**Drop columns**

We could drop *race*.

**Drop rows**

We can drop the NAs from *death_rate* and *age_adj_death_rate*

**Sort**

We can sort by *year* and *deaths*. That way we can see groupings per year of the highest cause of deaths.

**Filter** 

We can filter by sex, so to view only for men. 

**String manipulation**

We can `upper()` all values in sex. 

**Is it numeric?**

We can check if *cause* has numeric values. 

**Group and summary statistics** 

We can group by *cause* and get the summary statistics for it. 

**Group by two columns then sort within groups**

We can group by *year* and *cause*.

## Questions

After performing all those changes, we can ask the question "Per year, what cause has the highest amount of deaths" and we can also do that for death_rate! 

In [29]:
## Data Wrangling

import pandas as pd

cat_cols = ["year", "leading_cause", "sex", "race_ethnicity"]
num_cols = ["deaths", "death_rate", "age_adjusted_death_rate"]

# Rename columns
death_df = death_df.rename(columns={
    "leading_cause": "cause",
    "race_ethnicity": "race",
    "age_adjusted_death_rate": "age_adj_death_rate"
})

# Update column lists after renaming
cat_cols = ["year", "cause", "sex", "race"]
num_cols = ["deaths", "death_rate", "age_adj_death_rate"]

# Standardize sex values
death_df["sex"] = (
    death_df["sex"]
    .str.strip()
    .replace({
        "M": "Male",
        "F": "Female"
    })
)

# Create new column: difference between age-adjusted and crude death rate
death_df["adj_rate"] = (
    death_df["age_adj_death_rate"] - death_df["death_rate"]
)

# Drop rows with missing rate values
death_df = death_df.dropna(
    subset=["death_rate", "age_adj_death_rate"]
)

# Optional: drop race column
# death_df = death_df.drop(columns=["race"])

# Sort by year and deaths
death_df = death_df.sort_values(
    ["year", "deaths"],
    ascending=[True, False]
)

# Filter: only male records
male_df = death_df[death_df["sex"] == "Male"]

# Check if cause is numeric
cause_is_numeric = pd.api.types.is_numeric_dtype(death_df["cause"])
print("Is cause numeric?", cause_is_numeric)

# Group and summary statistics by cause
cause_summary = death_df.groupby("cause")["deaths"].describe()

# Group by year and cause, then sort within year
year_cause_deaths = (
    death_df
    .groupby(["year", "cause"], as_index=False)["deaths"]
    .sum()
    .sort_values(["year", "deaths"], ascending=[True, False])
)

display(death_df.head())
display(cause_summary)
display(year_cause_deaths.head(20))

Is cause numeric? False


,year,cause,sex,race_eth,deaths,death_rate,age_adj_death_rate,adj_rate
876,2015,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",Female,Non-Hispanic White,4494,317.960750,161.771435,-156.189314
949,2015,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",Male,Non-Hispanic White,4015,298.452877,239.223806,-59.229071
874,2015,Malignant Neoplasms (Cancer: C00-C97),Female,Non-Hispanic White,3130,221.454639,149.887787,-71.566851
952,2015,Malignant Neoplasms (Cancer: C00-C97),Male,Non-Hispanic White,3020,224.490084,186.128204,-38.361881
878,2015,All Other Causes,Female,Non-Hispanic White,2489,176.102427,108.574839,-67.527588


,count,mean,std,min,25%,50%,75%,max
cause,,,,,,,,
"Accidents Except Drug Poisoning (V01-X39, X43, X45-X59, Y85-Y86)",47.0,136.595745,83.610032,9.0,56.00,153.0,195.50,290.0
All Other Causes,60.0,1382.666667,786.489059,64.0,560.50,1415.5,2009.50,2777.0
Alzheimer's Disease (G30),34.0,172.264706,118.867206,6.0,59.50,165.0,229.25,378.0
"Assault (Homicide: U01-U02, Y87.1, X85-Y09)",6.0,196.833333,70.049744,133.0,153.00,162.0,252.75,291.0
"Assault (Homicide: Y87.1, X85-Y09)",1.0,196.000000,NaN,196.0,196.00,196.0,196.00,196.0
Cerebrovascular Disease (Stroke: I60-I69),59.0,227.474576,110.765262,7.0,123.50,228.0,303.00,451.0
Certain Conditions originating in the Perinatal Period (P00-P96),4.0,13.000000,4.546061,7.0,11.50,13.5,15.00,18.0
"Chronic Liver Disease and Cirrhosis (K70, K73)",1.0,166.000000,NaN,166.0,166.00,166.0,166.00,166.0
"Chronic Liver Disease and Cirrhosis (K70, K73-K74)",11.0,108.636364,74.894957,6.0,36.50,146.0,164.50,199.0


,year,cause,deaths
8,2015,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",16765
13,2015,Malignant Neoplasms (Cancer: C00-C97),13096
1,2015,All Other Causes,10973
11,2015,Influenza (Flu) and Pneumonia (J09-J18),2057
4,2015,Cerebrovascular Disease (Stroke: I60-I69),1811
7,2015,Diabetes Mellitus (E10-E14),1799
6,2015,Chronic Lower Respiratory Diseases (J40-J47),1720
9,2015,Essential Hypertension and Renal Diseases (I10...,1078
2,2015,Alzheimer's Disease (G30),921
14,2015,Mental and Behavioral Disorders due to Acciden...,836


In [31]:
# Question

# Per year, cause with the highest total deaths
top_deaths_per_year = (
    death_df
    .groupby(["year", "cause"], as_index=False)["deaths"]
    .sum()
    .loc[lambda df: df.groupby("year")["deaths"].idxmax()]
    .sort_values("year")
)

display(top_deaths_per_year)

# Per year, cause with the highest average death rate
top_death_rate_per_year = (
    death_df
    .groupby(["year", "cause"], as_index=False)["death_rate"]
    .mean()
    .loc[lambda df: df.groupby("year")["death_rate"].idxmax()]
    .sort_values("year")
)

display(top_death_rate_per_year)

,year,cause,deaths
8,2015,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",16765
25,2016,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",16805
40,2017,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",17121
56,2018,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",17606
73,2019,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",17667
90,2020,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",20606
105,2021,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",16074


,year,cause,death_rate
8,2015,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",184.995148
25,2016,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",186.645173
40,2017,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",190.240828
56,2018,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",196.011700
73,2019,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",197.037067
88,2020,Covid-19,236.162500
105,2021,"Diseases of Heart (I00-I09, I11, I13, I20-I51)",186.437500


# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?